typed-protocols/src/Network/TypedProtocol/Driver/ByteLimit.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE PolyKinds           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Network.TypedProtocol.Driver.ByteLimit
+  ( runPeerWithByteLimit
+  , DecoderFailureOrTooMuchInput (..)
+  ) where
+
+import           Control.Exception
+import           Data.Int (Int64)
+
+import           Control.Monad.Class.MonadThrow (MonadThrow (..))
+import           Control.Tracer (Tracer, traceWith)
+
+
+import           Network.TypedProtocol.Core
+import           Network.TypedProtocol.Codec
+import           Network.TypedProtocol.Channel
+import           Network.TypedProtocol.Driver
+
+
+data DecoderFailureOrTooMuchInput failure
+  = DecoderFailure !failure
+  | TooMuchInput
+  deriving (Show)
+
+instance Exception failure => Exception (DecoderFailureOrTooMuchInput failure)
+
+runDecoderWithByteLimit
+    :: forall m bytes failure a. Monad m
+    => Int64
+    -- ^ message size limit
+    -> (bytes -> Int64)
+    -- ^ byte size
+    -> Channel m bytes
+    -> Maybe bytes
+    -> DecodeStep bytes failure m a
+    -> m (Either (DecoderFailureOrTooMuchInput failure) (a, Maybe bytes))
+runDecoderWithByteLimit limit byteLength Channel{recv} mbytes = go (limit - maybe 0 byteLength mbytes) mbytes
+    where
+      go :: Int64
+         -> Maybe bytes
+         -> DecodeStep bytes failure m a
+         -> m (Either (DecoderFailureOrTooMuchInput failure) (a, Maybe bytes))
+      -- we run over the limit, return @TooMuchInput@ error
+      go !l _  _                         | l < 0 = return (Left TooMuchInput)
```

Comment:
When it's `DecodeDone x trailing` where `mbytes = Just bytes`, `trailing = tail bytes`, `length trailing > limit`, it shouldn't fail, no? The decoder only used one byte of the trailing, yet the trailing bytes tripped the limit.

Reply:
So is the limit about how much the decoder actually consumed? Or about how much input is made available to the decoder?

Reply:
The limit is about how much the decoder actually consumed.  The trailing part will count for the next message.

Reply:
Then it's not correct, is it? `runDecoderWithByteLimit limit byteLength channel (Just bytes) decoder` will fail whenever `byteLength bytes > limit`, even if the decoder only uses a part of `bytes` which is shorter than `limit`.

Reply:
Yes, you're right.

Reply:
I fixed this, and I added a quickcheck property `prop_runPeerWithByteLimit` which checks it.